modules/rpc/yarpc.go

```diff
@@ -36,66 +35,138 @@ import (
+type registerServiceFunc func(module *YARPCModule)
 
 type yarpcConfig struct {
 	modules.ModuleConfig
-	Bind          string `yaml:"bind"`
-	AdvertiseName string `yaml:"advertiseName"`
+	Bind                    string `yaml:"bind"`
+	AdvertiseName           string `yaml:"advertiseName"`
+	inboundMiddleware       []middleware.UnaryInbound
+	onewayInboundMiddleware []middleware.OnewayInbound
+	inbounds                []transport.Inbound
+}
+
+// Stores a collection of all modules configs and provides
+// operations that are safe to call from multiple go routines.
+type configCollection struct {
+	sync.RWMutex
+	configs []*yarpcConfig
+}
+
+func (c *configCollection) addConfig(config *yarpcConfig) error {
+	c.Lock()
+	defer c.Unlock()
+
+	if len(c.configs) > 0 && config.AdvertiseName != c.configs[0].AdvertiseName {
+		return errors.New("Name mismatch")
+	}
+
+	c.configs = append(c.configs, config)
+	return nil
+}
+
+func (c *configCollection) removeConfig(config *yarpcConfig) error {
+	c.Lock()
+	defer c.Unlock()
+
+	for i := range c.configs {
+		if c.configs[i] == config {
+			c.configs = append(c.configs[:i], c.configs[i+1:]...)
+			return nil
+		}
+	}
+
+	return errors.New("config not found")
+}
+
+func (c *configCollection) mergeConfigs() (conf yarpc.Config, err error) {
+	c.RLock()
+	defer c.RUnlock()
+
+	if len(c.configs) <= 1 {
+		err = errors.New("empty configs")
+		return
+	}
+
+	conf.Name = c.configs[0].AdvertiseName
+	var inboundMiddleware []middleware.UnaryInbound
+	var onewayInboundMiddleware []middleware.OnewayInbound
+	for _, cfg := range c.configs {
+		conf.Inbounds = append(conf.Inbounds, cfg.inbounds...)
+		inboundMiddleware = append(inboundMiddleware, cfg.inboundMiddleware...)
+		onewayInboundMiddleware = append(onewayInboundMiddleware, cfg.onewayInboundMiddleware...)
+	}
+
+	conf.InboundMiddleware = yarpc.InboundMiddleware{
+		Unary:  yarpc.UnaryInboundMiddleware(inboundMiddleware...),
+		Oneway: yarpc.OnewayInboundMiddleware(onewayInboundMiddleware...),
+	}
+
+	return conf, nil
 }
 
-func newYarpcModule(
+func newYARPCModule(
 	mi service.ModuleCreateInfo,
 	reg registerServiceFunc,
 	options ...modules.Option,
-) (*YarpcModule, error) {
-	cfg := &yarpcConfig{
-		AdvertiseName: mi.Host.Name(),
-		Bind:          ":0",
-	}
-
+) (*YARPCModule, error) {
 	name := "yarpc"
 	if mi.Name != "" {
 		name = mi.Name
 	}
 
-	module := &YarpcModule{
+	module := &YARPCModule{
 		ModuleBase: *modules.NewModuleBase(name, mi.Host, []string{}),
 		register:   reg,
-		config:     *cfg,
-	}
-
-	options = append([]modules.Option{WithInboundMiddleware(fxContextInboundMiddleware{
-		Host: mi.Host,
-	}),
-		WithInboundMiddleware(authInboundMiddleware{
-			Host: mi.Host,
-		}),
-		WithOnewayInboundMiddleware(fxContextOnewayInboundMiddleware{
-			Host: mi.Host,
-		}),
-		WithOnewayInboundMiddleware(authOnewayInboundMiddleware{
-			Host: mi.Host,
-		}),
-	}, options...)
+		config: yarpcConfig{
+			AdvertiseName: mi.Host.Name(),
```

Comment:
I was confused about it as well: this is a default value, that is going to be overridden by the PopulateStruct, so here we simply remove a default port ":0" if it is not in the config.

I am open to discussion if we want to keep it as default though.